engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnMessageBehavior.java

```diff
@@ -0,0 +1,173 @@
+              writeMessagePublishCommand(properties, variables);
+              return null;
+            });
+  }
+
+  private Either<Failure, MessageProperties> evaluateMessageExpressions(
+      final PublishMessageProperties publishMessageProps, final long scopeKey) {
+    return Either.<Failure, MessageProperties>right(new MessageProperties())
+        .flatMap(p -> evalMessageNameExp(publishMessageProps, scopeKey).map(p::messageName))
+        .flatMap(p -> evalMessageIdExp(publishMessageProps, scopeKey).map(p::messageId))
+        .flatMap(p -> evalCorrelationKeyExp(publishMessageProps, scopeKey).map(p::correlationKey))
+        .flatMap(p -> evalTimeToLiveExp(publishMessageProps, scopeKey).map(p::timeToLive));
+  }
+
+  private Either<Failure, String> evalMessageNameExp(
+      final PublishMessageProperties publishMessageProps, final long scopeKey) {
+    return expressionBehavior.evaluateStringExpression(
+        publishMessageProps.getMessageName(), scopeKey);
+  }
+
+  private Either<Failure, String> evalMessageIdExp(
+      final PublishMessageProperties publishMessageProps, final long scopeKey) {
+    final Expression messageId = publishMessageProps.getMessageId();
+    if (messageId == null) {
+      return Either.right(null);
+    }
+    return expressionBehavior.evaluateStringExpression(
+        publishMessageProps.getMessageId(), scopeKey);
+  }
+
+  private Either<Failure, String> evalCorrelationKeyExp(
+      final PublishMessageProperties publishMessageProps, final long scopeKey) {
+    return expressionBehavior.evaluateMessageCorrelationKeyExpression(
+        publishMessageProps.getCorrelationKey(), scopeKey);
+  }
+
+  private Either<Failure, Long> evalTimeToLiveExp(
+      final PublishMessageProperties publishMessageProps, final long scopeKey) {
+    final Expression timeToLive = publishMessageProps.getTimeToLive();
+    if (timeToLive == null) {
+      return Either.right(null);
+    }
+
+    return expressionBehavior
+        .evaluateIntervalExpression(timeToLive, scopeKey)
+        .map(
+            interval -> {
+              final var currentTimeMillis = ActorClock.currentTimeMillis();
+              return interval.toEpochMilli(currentTimeMillis) - currentTimeMillis;
+            });
+  }
+
+  private void writeMessagePublishCommand(
+      final MessageProperties properties, final DirectBuffer variables) {
+
+    messageRecord.reset();
+    messageRecord.setName(properties.getMessageName());
+    messageRecord.setCorrelationKey(properties.getCorrelationKey());
+    messageRecord.setVariables(variables);
+    Optional.ofNullable(properties.getMessageId()).ifPresent(messageRecord::setMessageId);
+
+    final var timeToLive =
+        Optional.ofNullable(properties.getTimeToLive()).orElse(DEFAULT_MSG_TTL.toMillis());
+    messageRecord.setTimeToLive(timeToLive);
+
+    final var key = keyGenerator.nextKey();
+    commandWriter.appendFollowUpCommand(key, MessageIntent.PUBLISH, messageRecord);
```

Comment:
❌ The message must be published on a specific partition, and without a key.

The partition id is selected by hashing the `correlationKey` to the partition where a subscription for this specific message can be opened.

💡 You can use `SubscriptionUtil.getSubscriptionPartitionId` to find the correct partition id.

https://github.com/camunda/zeebe/blob/1c84f9694d7be612f6d346e436d1c6c25ab160dc/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/SubscriptionUtil.java#L32-L44

💡 For sending the message to the selected partition, you can use `InterPartitionCommandSender.sendCommand` (choose the option that does not take a `recordKey`).

https://github.com/camunda/zeebe/blob/1c84f9694d7be612f6d346e436d1c6c25ab160dc/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java#L14-L23